test/Web/ScottySpec.hs

```diff
@@ -36,176 +39,181 @@ availableMethods = [GET, POST, HEAD, PUT, PATCH, DELETE, OPTIONS]
+                    request
+                        "POST"
+                        "/upload"
+                        [("Content-Type", "multipart/form-data; boundary=--33")]
+                        (TLE.encodeUtf8 . TL.pack . concat $ [show c | c <- ([1 .. 50] :: [Integer])])
+                        `shouldRespondWith` 200
+
+        describe "text" $ do
+            let modernGreekText :: IsString a => a
+                modernGreekText = "νέα ελληνικά"
+
+            withApp (Scotty.get "/scotty" $ text modernGreekText) $ do
+                it "sets body to given text" $ do
+                    get "/scotty" `shouldRespondWith` modernGreekText
+
+                it "sets Content-Type header to \"text/plain; charset=utf-8\"" $ do
+                    get "/scotty" `shouldRespondWith` 200{matchHeaders = ["Content-Type" <:> "text/plain; charset=utf-8"]}
+
+            withApp (Scotty.get "/scotty" $ setHeader "Content-Type" "text/somethingweird" >> text modernGreekText) $ do
+                it "doesn't override a previously set Content-Type header" $ do
+                    get "/scotty" `shouldRespondWith` 200{matchHeaders = ["Content-Type" <:> "text/somethingweird"]}
+
+        describe "html" $ do
+            let russianLanguageTextInHtml :: IsString a => a
+                russianLanguageTextInHtml = "<p>ру́сский язы́к</p>"
+
+            withApp (Scotty.get "/scotty" $ html russianLanguageTextInHtml) $ do
+                it "sets body to given text" $ do
+                    get "/scotty" `shouldRespondWith` russianLanguageTextInHtml
+
+                it "sets Content-Type header to \"text/html; charset=utf-8\"" $ do
+                    get "/scotty" `shouldRespondWith` 200{matchHeaders = ["Content-Type" <:> "text/html; charset=utf-8"]}
+
+            withApp (Scotty.get "/scotty" $ setHeader "Content-Type" "text/somethingweird" >> html russianLanguageTextInHtml) $ do
+                it "doesn't override a previously set Content-Type header" $ do
+                    get "/scotty" `shouldRespondWith` 200{matchHeaders = ["Content-Type" <:> "text/somethingweird"]}
+
+        describe "json" $ do
+            withApp (Scotty.get "/scotty" $ setHeader "Content-Type" "text/somethingweird" >> json (Just (5 :: Int))) $ do
+                it "doesn't override a previously set Content-Type header" $ do
+                    get "/scotty" `shouldRespondWith` 200{matchHeaders = ["Content-Type" <:> "text/somethingweird"]}
+
+        describe "finish" $ do
+            withApp (Scotty.get "/scotty" finish) $ do
+                it "responds with 200 by default" $ do
+                    get "/scotty" `shouldRespondWith` 200
+
+            withApp (Scotty.get "/scotty" $ status status400 >> finish >> status status200) $ do
+                it "stops the execution of an action" $ do
+                    get "/scotty" `shouldRespondWith` 400
+
+-- Skip these tests on Windows, since Unix sockets are not available
 #if !defined(mingw32_HOST_OS)
```

Comment:
I moved some things that were top-level and nested them into the test block. Reads better with the CPP that way.